server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1631,63 +1636,93 @@ void removeEntitlement(Entitlement entitlement, boolean regenModified) {
         }
 
         // post unbind actions
-        PoolHelper poolHelper = new PoolHelper(this, entitlement);
-        enforcer.postUnbind(consumer, poolHelper, entitlement);
+        for (Entitlement ent : entsToRevoke) {
+            PoolHelper poolHelper = new PoolHelper(this,  ent);
+            enforcer.postUnbind(ent.getConsumer(), poolHelper, ent);
+        }
+
+        List<Entitlement> batch = new ArrayList<Entitlement>();
+        for (int i = 0; i < entsToRevoke.size(); i++) {
+            Entitlement entitlement = entsToRevoke.get(i);
+            batch.add(entitlement);
 
-        if (regenModified) {
-            // Find all of the entitlements that modified the original entitlement,
-            // and regenerate those to remove the content sets.
-            // Lazy regeneration is ok here.
-            this.regenerateCertificatesOf(entitlementCurator.listModifying(entitlement), true);
+            // We work in batches of maximum size 1000.
+            if (i % 1000 == 0) {
+                Set<Entitlement> modifiedEnts = entitlementCurator.batchListModifying(batch);
+                this.regenerateCertificatesOf(modifiedEnts, true);
+                batch.clear();
```

Comment:
No the reason is not _in clause_ limitation. The reason is to optimize underlying batchListModifying method. The method batchListModifying starts by searching overlapping product ids:
```java
ProductEntitlements pidEnts = getOverlappingForModifying(entitlements);
```
The getOverlappingForModifying might return high number of entitlements. More then actually necessary. The rest of the batchListModifying method filters pidEnts. So this batch size (1000) actually gives us a way to configure the following trade off: 

If the _batch size_ is too big, the batchListModifying starts by finding too many entitlements and then consumes a lot of memory. On the other hand if _batch size_ is too small, we end up with too many calls to batchListModifying which is better on memory but worse on time performance.

However, thinking about this a bit more, I agree the batch size of 1000 is kinda arbitrary. It worked well on the problematic Import dataset. But again I agree its really arbitrary constant :-P